torch/_inductor/codegen/triton.py

```diff
@@ -3094,26 +3094,88 @@ def get_load_buffer(self, indexing):
         else:
             return self.loads
 
-    def _handle_pdl_before_load(self, wait_buffer):
+    def _enable_pdl_codegen(self):
+        if not torch._inductor.config.triton.enable_pdl:
+            return False
+        if isinstance(V.kernel, torch._inductor.select_algorithm.TritonTemplateKernel):
+            return False
+        return (
+            V.graph.get_current_device_or_throw().type == "cuda"
+            and torch.cuda.get_device_capability()[0] >= 9
+        )
+
+    def _handle_pdl_before_access(
+        self, wait_buffer, *dependencies, consider_reads=False
+    ):
+        if not self._enable_pdl_codegen():
+            return
+        current_node = V.kernel.current_node
+        prev_node = (
+            V.graph.scheduler.previous_node if V.graph.scheduler is not None else None
+        )
+
+        def matching_dep(dep):
+            assert prev_node is not None
+            prev_deps = prev_node.read_writes.writes
+            if consider_reads:
+                prev_deps = itertools.chain(prev_deps, prev_node.read_writes.reads)
+            return any(
+                dep == current_node.mutation_renames.get(w.name, w.name)
+                for w in prev_deps
+            )
+
+        assert dependencies
+        need_wait = (
+            prev_node is None
+            or isinstance(prev_node, NopKernelSchedulerNode)
```

Comment:
Would it make more sense to set prev_node = None within NopKernelSchedulerNode.

Reply:
good question - so far i like having prev_node being managed in only one place, i feel like that is more clear.

but i agree that special-casing node types is a little clunky, and it would be better to have some sort of property that gets queried when we move up prev_node instead. i guess (also to the end of cond/while/subgraph) the more conservative option would be to let scheduler nodes opt in rather for them to opt out. this might also be a good way to eventually handle streams.

does that make sense?